File: frontend/src/models/item/purchase/payment-instrument.ts
import { PaymentInstrumentDTO } from '../../../data-services/api-services/models';
import { Address } from './address';

export class PaymentInstrument {
    type: string = '';

    email: string = '';

    purchaseAmount: number = 0;

    billingAddress: Address;

    merchantPaymentInstrumentId: string = '';

    creationDate: string = '';

    updateDate: string = '';

    state: string = '';

    cardType: string = '';

    holderName: string = '';

    BIN: string = '';

    expirationDate: string = '';

    lastFourDigits: string = '';

    billingAgreementId: string = '';

    payerId: string = '';

    payerStatus: string = '';

    addressStatus: string = '';

    IMEI: string = '';

    merchantLocalDate: string = '';

    paymentInstrumentId: string = '';

    purchaseAmountInUSD: number = 0;

    constructor() {
        this.billingAddress = new Address();
    }

    fromDTO(paymentInstrument: PaymentInstrumentDTO) {
        const {
            Type,
            Email,
            PurchaseAmount,
            BillingAddress: billingAddress,
            MerchantPaymentInstrumentId,
            CreationDate,
            UpdateDate,
            State,
            CardType,
            HolderName,
            BIN,
            ExpirationDate,
            LastFourDigits,
            BillingAgreementId,
            PayerId,
            PayerStatus,
            AddressStatus,
            IMEI,
            MerchantLocalDate,
            PaymentInstrumentId,
            PurchaseAmountInUSD,
        } = paymentInstrument;

        this.type = Type;
        this.email = Email;
        this.purchaseAmount = PurchaseAmount;
        if (billingAddress) {
            this.billingAddress.fromDTO(billingAddress);
        }
        this.merchantPaymentInstrumentId = MerchantPaymentInstrumentId;
        this.creationDate = CreationDate;
        this.updateDate = UpdateDate;
        this.state = State;
        this.cardType = CardType;
        this.holderName = HolderName;
        this.BIN = BIN;
        this.expirationDate = ExpirationDate;
        this.lastFourDigits = LastFourDigits;
        this.billingAgreementId = BillingAgreementId;
        this.payerId = PayerId;
        this.payerStatus = PayerStatus;
        this.addressStatus = AddressStatus;
        this.IMEI = IMEI;
        this.merchantLocalDate = MerchantLocalDate;
        this.paymentInstrumentId = PaymentInstrumentId;
        this.purchaseAmountInUSD = PurchaseAmountInUSD;

        return this;
    }
}

File: backend/model/src/main/java/com/griddynamics/msd365fp/manualreview/model/dfp/PaymentInstrument.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.griddynamics.msd365fp.manualreview.model.dfp;

import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.griddynamics.msd365fp.manualreview.model.jackson.EpochSecondsDateTimeSerializer;
import com.griddynamics.msd365fp.manualreview.model.jackson.FlexibleDateFormatDeserializer;
import com.griddynamics.msd365fp.manualreview.model.jackson.ISOStringDateTimeSerializer;
import lombok.Data;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.HashMap;
import java.util.Map;

@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.UpperCamelCaseStrategy.class)
public class PaymentInstrument implements Serializable {
    private String paymentInstrumentId;
    private BigDecimal purchaseAmount;
    private BigDecimal purchaseAmountInUSD;
    private String merchantPaymentInstrumentId;
    private String type;
    @JsonDeserialize(using = FlexibleDateFormatDeserializer.class)
    @JsonSerialize(using = ISOStringDateTimeSerializer.class)
    private OffsetDateTime creationDate;
    @JsonDeserialize(using = FlexibleDateFormatDeserializer.class)
    @JsonSerialize(using = ISOStringDateTimeSerializer.class)
    private OffsetDateTime updateDate;
    private String state;
    private String cardType;
    private String holderName;
    @JsonProperty("BIN")
    private String bin;
    private String expirationDate;
    private String lastFourDigits;
    private String email;
    private String billingAgreementId;
    private String payerId;
    private String payerStatus;
    private String addressStatus;
    @JsonProperty("IMEI")
    private String imei;
    private String addressId;
    @JsonDeserialize(using = FlexibleDateFormatDeserializer.class)
    @JsonSerialize(using = ISOStringDateTimeSerializer.class)
    private OffsetDateTime merchantLocalDate;

    private Map<String, String> additionalParams = new HashMap<>();

    @JsonAnySetter
    public void setAdditionalParam(String name, String value) {
        additionalParams.put(name, value);
    }
    public void setAdditionalParams(Map<String, String> map) {
        additionalParams.putAll(map);
    }

    /**
     * Getter for advanced serialization.
     * Json object will contain both representation of DateTime field -
     * timestamp (for SQL queries) and string representation.
     */
    @SuppressWarnings("unused")
    @JsonSerialize(using = EpochSecondsDateTimeSerializer.class)
    public OffsetDateTime getCreationDateEpochSeconds() {
        return creationDate;
    }

    /**
     * Getter for advanced serialization.
     * Json object will contain both representation of DateTime field -
     * timestamp (for SQL queries) and string representation.
     */
    @SuppressWarnings("unused")
    @JsonSerialize(using = EpochSecondsDateTimeSerializer.class)
    public OffsetDateTime getUpdateDateEpochSeconds() {
        return updateDate;
    }

    /**
     * Getter for advanced serialization.
     * Json object will contain both representation of DateTime field -
     * timestamp (for SQL queries) and string representation.
     */
    @SuppressWarnings("unused")
    @JsonSerialize(using = EpochSecondsDateTimeSerializer.class)
    public OffsetDateTime getMerchantLocalDateEpochSeconds() {
        return merchantLocalDate;
    }
}
